// gzipper/dist/bin/index.js
#!/usr/bin/env node
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var commander_1 = __importDefault(require("commander"));
var Gzipper_1 = require("../src/Gzipper");
var version = require('../package.json').version;
var Index = /** @class */ (function () {
    function Index() {
        this.argv = process.argv;
        this.env = process.env;
    }
    Index.prototype.getOptions = function () {
        commander_1.default
            .version(version)
            .usage('[options] <path> [outputPath]')
            .option('-v, --verbose', 'detailed level of logs')
            .option('-e, --exclude <extensions>', 'exclude file extensions from compression, example: jpeg,jpg...', this.optionToArray.bind(this))
            .option('-i, --include <extensions>', 'include file extensions for compression, example: js,css,html...', this.optionToArray.bind(this))
            .option('-t, --threshold <number>', 'exclude assets smaller than this byte size. 0 (default)', function (value) { return parseInt(value); })
            // TODO: Remove @deprecated --gzip-level
            .option('--gzip-level <number> (deprecated)', 'gzip compression level 6 (default), 0 (no compression) - 9 (best compression)', function (value) { return parseInt(value); })
            .option('--level <number>', 'compression level 6 (default), 0 (no compression) - 9 (best compression)', function (value) { return parseInt(value); })
            // TODO: Remove @deprecated --gzip-memory-level
            .option('--gzip-memory-level <number> (deprecated)', 'amount of memory which will be allocated for compression 8 (default), 1 (minimum memory) - 9 (maximum memory)', function (value) { return parseInt(value); })
            .option('--memory-level <number>', 'amount of memory which will be allocated for compression 8 (default), 1 (minimum memory) - 9 (maximum memory)', function (value) { return parseInt(value); })
            // TODO: Remove @deprecated --gzip-strategy
            .option('--gzip-strategy <number> (deprecated)', 'compression strategy 0 (default), 1 (filtered), 2 (huffman only), 3 (RLE), 4 (fixed)', function (value) { return parseInt(value); })
            .option('--strategy <number>', 'compression strategy 0 (default), 1 (filtered), 2 (huffman only), 3 (RLE), 4 (fixed)', function (value) { return parseInt(value); })
            .option('--deflate', 'enable deflate compression')
            .option('--brotli', 'enable brotli compression, Node.js >= v11.7.0')
            .option('--brotli-param-mode <value>', 'default, text (for UTF-8 text), font (for WOFF 2.0 fonts)')
            .option('--brotli-quality <number>', 'brotli compression quality 11 (default), 0 - 11', function (value) { return parseInt(value); })
            .option('--brotli-size-hint <number>', 'expected input size 0 (default)', function (value) { return parseInt(value); })
            .option('--output-file-format <value>', 'output file format with default artifacts [filename].[ext].[compressExt]')
            .option('', 'where:')
            .option('', 'filename -> file name')
            .option('', 'ext -> file extension')
            .option('', 'compressExt -> compress extension (.gz, .br, etc)')
            .option('', 'hash -> uniq uuid/v4 hash')
            .option('', 'samples:')
            .option('', '[filename].[compressExt].[ext]')
            .option('', 'test-[filename]-[hash].[compressExt].[ext]')
            .option('', '[filename]-[hash]-[filename]-tmp.[ext].[compressExt]')
            .parse(this.argv)
            .removeAllListeners();
        var _a = commander_1.default.args, target = _a[0], outputPath = _a[1];
        var options = {
            verbose: this.env.GZIPPER_VERBOSE
                ? !!parseInt(this.env.GZIPPER_VERBOSE)
                : commander_1.default.verbose,
            exclude: this.optionToArray(this.env.GZIPPER_EXCLUDE) ||
                commander_1.default.exclude,
            include: this.optionToArray(this.env.GZIPPER_INCLUDE) ||
                commander_1.default.include,
            threshold: parseInt(this.env.GZIPPER_THRESHOLD) ||
                commander_1.default.threshold ||
                0,
            // TODO: Remove @deprecated GZIPPER_GZIP_LEVEL, gzipLevel
            level: parseInt(this.env.GZIPPER_LEVEL) ||
                commander_1.default.level ||
                parseInt(this.env.GZIPPER_GZIP_LEVEL) ||
                commander_1.default.gzipLevel,
            // TODO: Remove @deprecated GZIPPER_GZIP_MEMORY_LEVEL, gzipMemoryLevel
            memoryLevel: parseInt(this.env.GZIPPER_MEMORY_LEVEL) ||
                commander_1.default.memoryLevel ||
                parseInt(this.env.GZIPPER_GZIP_MEMORY_LEVEL) ||
                commander_1.default.gzipMemoryLevel,
            // TODO: Remove @deprecated GZIPPER_GZIP_STRATEGY, gzipStrategy
            strategy: parseInt(this.env.GZIPPER_STRATEGY) ||
                commander_1.default.strategy ||
                parseInt(this.env.GZIPPER_GZIP_STRATEGY) ||
                commander_1.default.gzipStrategy,
            brotli: this.env.GZIPPER_BROTLI
                ? !!parseInt(this.env.GZIPPER_BROTLI)
                : commander_1.default.brotli,
            deflate: this.env.GZIPPER_DEFLATE
                ? !!parseInt(this.env.GZIPPER_DEFLATE)
                : commander_1.default.deflate,
            brotliParamMode: this.env.GZIPPER_BROTLI_PARAM_MODE || commander_1.default.brotliParamMode,
            brotliQuality: parseInt(this.env.GZIPPER_BROTLI_QUALITY) ||
                commander_1.default.brotliQuality,
            brotliSizeHint: parseInt(this.env.GZIPPER_BROTLI_SIZE_HINT) ||
                commander_1.default.brotliSizeHint,
            outputFileFormat: this.env.GZIPPER_OUTPUT_FILE_FORMAT || commander_1.default.outputFileFormat,
        };
        this.target = target;
        this.outputPath = outputPath;
        this.options = options;
        return this;
    };
    // Delete undefined and NaN options.
    Index.prototype.filterOptions = function () {
        var _this = this;
        Object.keys(this.options).forEach(function (key) {
            if (_this.options[key] === undefined ||
                _this.options[key] !==
                    _this.options[key]) {
                delete _this.options[key];
            }
        });
        return this;
    };
    Index.prototype.start = function () {
        var gzipper = new Gzipper_1.Gzipper(this.target, this.outputPath, this.options || {});
        gzipper.compress().catch(function (err) { return console.error(err); });
    };
    Index.prototype.optionToArray = function (value) {
        if (value) {
            return value.split(',').map(function (item) { return "." + item.trim(); });
        }
        return value;
    };
    return Index;
}());
exports.Index = Index;
if (process.env.NODE_ENV !== 'testing') {
    new Index()
        .getOptions()
        .filterOptions()
        .start();
}
