// gzipper/dist/src/Logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Custom logger.
 */
var Logger = /** @class */ (function () {
    /**
     * Creates an instance of Logger.
     */
    function Logger(verbose) {
        this.verbose = verbose;
    }
    /**
     * Info logger.
     */
    Logger.prototype.info = function (message, force) {
        return this.logger('info')(message, force);
    };
    /**
     * Error logger.
     */
    Logger.prototype.error = function (message, force) {
        return this.logger('error')(message, force);
    };
    /**
     * Success logger.
     */
    Logger.prototype.success = function (message, force) {
        return this.logger('success')(message, force);
    };
    /**
     * Warn logger.
     */
    Logger.prototype.warn = function (message, force) {
        return this.logger('warning')(message, force);
    };
    /**
     * Colorize messages depends on the level and return a wrapper.
     */
    Logger.prototype.logger = function (level) {
        var _this = this;
        var colorfulMessage;
        var prefix = 'gzipper: ';
        switch (level) {
            case 'info':
                colorfulMessage = "\u001B[36m" + prefix + "%s\u001B[0m";
                break;
            case 'error':
                colorfulMessage = "\u001B[31m" + prefix + "%s\u001B[0m";
                break;
            case 'warning':
                colorfulMessage = "\u001B[33m" + prefix + "%s\u001B[0m";
                break;
            case 'success':
                colorfulMessage = "\u001B[32m" + prefix + "%s\u001B[0m";
                break;
        }
        return function (message, force) {
            if (_this.verbose || force) {
                console.log(colorfulMessage, message);
            }
        };
    };
    return Logger;
}());
exports.Logger = Logger;
